### ISSUE TITLE
Injection of Localizables not working in @Pages

### ISSUE THREAD

**Issue by RickStrahl** [aspnet/Localization#420]
I'm trying to get localization to work inside of a test Razor Page, but really nothing is working to inject localizers or use View localization. 

Injecting IViewLocalizer I get an IndexOutofRange Exception as soon as:

```html
@inject IViewLocalizer viewLocalizer
```
is added to the page (even if nothing uses it). Page runs fine without this injected value. 

Error is:

![image](https://user-images.githubusercontent.com/1374013/31485578-0bd31b66-aed0-11e7-96e7-191cd628f813.png)

It was suggested by Damien to use IStringLocalizer instead. So I tried that and that also seems to not work. I get no errors but when I inject:

```
@inject IStringLocalizer<AspNetCoreLocalization.Properties.CommonResources> stringLocalizer
```

The localizer is initialized but the ResourceManager contains no resources. 

I can use the same lStringLocalizer either inside of an MVC controller or an MVC View and it works, but inside of a Page I get nothing.

The usage is:

```
@stringLocalizer["HelloWorld"]
```

and it produces - **empty output**. Not the resource value, not the key value either - just empty.

Pretty sure the code is set up right and the right resources exist since it's working in controller and View, but just in case here's my set up code (which is all straight out of the docs):

```cs
 public void ConfigureServices(IServiceCollection services)
 {
            services
                .AddLocalization(options => options.ResourcesPath = "Properties");

            services.AddMvc()                
                .AddViewLocalization(LanguageViewLocationExpanderFormat.Suffix)
                .AddDataAnnotationsLocalization();
}
public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
     ...
            var cultures = new[]
            {
                new CultureInfo("en-US"),
                new CultureInfo("en"),
                new CultureInfo("de-DE"),
                new CultureInfo("de"),
                new CultureInfo("fr"),
                //new CultureInfo("zh"),
                //new CultureInfo("pl")
            };
            app.UseRequestLocalization(new RequestLocalizationOptions
            {
                DefaultRequestCulture = new RequestCulture("en"),
                SupportedCultures = cultures,
                SupportedUICultures = cultures
            });

            app.UseStaticFiles();

            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller}/{action=Index}/{id?}");
            });
        }
```

To summarize:

* Injecting IViewLocalizer in a @Page fails outright
* Injecting IStringLocalizer<T> into a @Page works, but produces no output

* Injecting the same IStringLocalizer<T> into a controller works
* Injecting IStringLocalizer<T> into an MVC view works
* Inject IViewLocalizer into an MVC View works

**Comment by RickStrahl** [comment]
After some more experimenting with a new, clean project I can confirm that IViewLocalizer definitely isn't working, but I think IStringLocalizer is actually working the way it's supposed to.

It looks like the IStringLocalizer<T> issue I was seeing was related to the an invalid resource path. 

If I have a 'placehold class' for StringLocalizer definied like this:

```
namespace AspNetCoreLocalization.Properties
{
    /// <summary>
    /// Classes exist just for providing a reference
    /// </summary>
    public class CommonResources
    {  }
}
````
so I can load  `IStringLocalizer<CommonResources>` I have to have a class defined as `Properties.CommonResources` which gets created as `AspNetLocalization.Properties.Properties.CommonResources. 

That extra injected `Properties`  is bloody confusing as hell and that's what caused me to not be able to access my string localizer resources.

Anyway after accounting for this wonkyness and declaring my resources properly IStringLocalizer worked.

As an aside this brings up another problem: 

When there are no matching resources StringLocalizer et al don't fail, but they also **return absolute no resources at all**. Not the resource key, but nothing. 

The correct behavior IMHO is to either fail if the resource set assigned by the type can't be matched or just like any other missing individual resource key, display the resource key.

**Comment by hishamco** [comment]
@RickStrahl I already reply to your tweet, there a bug on `IViewLocalizer` on ASP.NET 2.0.0, there's the workaround http://www.hishambinateya.com/localization-issue-in-razorpages

**Comment by hishamco** [comment]
FYI https://github.com/aspnet/Mvc/issues/6694

**Comment by DamianEdwards** [comment]
The exception is a known issue that's already fixed (aspnet/Mvc#6694) and I've provided an example that shows how to work around it locally at https://github.com/DamianEdwards/RazorPagesLocalization. That will let you keep using `IViewLocalizer` until we deploy the fix.

As for `StringLocalizer` behavior when the requested string can't be found, it **does** return the string passed as the key/name. Can you provide a more detailed repro of when it doesn't do that?

**Comment by hishamco** [comment]
I think this need to be closed, while it's fixed in https://github.com/aspnet/Mvc/pull/6838

**Comment by RickStrahl** [comment]
Thank you. I've been using Damiens fix and that worked for me. Haven't checked in 2.0.1 but I presume it was fixed for that?

**Comment by pranavkm** [comment]
@RickStrahl yup. The issue should be fixed in Mvc 2.0.1.
